lib/posix/spawn/child.rb

```diff
@@ -244,14 +251,35 @@ def read_and_write(input, stdin, stdout, stderr, timeout=nil, max=nil)
 
           # read from stdout and stderr streams
           ready[0].each do |fd|
-            buf = (fd == stdout) ? @out : @err
+            chunk = nil
             begin
-              buf << fd.readpartial(BUFSIZE)
+              chunk = fd.readpartial(BUFSIZE)
             rescue Errno::EAGAIN, Errno::EINTR
             rescue EOFError
               readers.delete(fd)
               fd.close
             end
+
+            abort = false
+            if chunk
+              if fd == stdout
+                if @streaming && @stdout_block
+                  abort = !!@stdout_block.call(chunk)
+                else
+                  @out << chunk
+                end
+              else
+                if @streaming && @stderr_block
+                  abort = !!@stderr_block.call(chunk)
+                else
+                  @err << chunk
+                end
+              end
+            end
```

Comment:
This might be the "tricky" Ruby you're talking about, but it seems to me that when streaming is not requested, you could set `@stdout_block` anyway, to

    Proc.new do |chunk|
      @out << chunk
      false
    end

(like you do for the tests) and do the equivalent for `@stderr_block`. Then you could avoid the inner conditionals here.

To shrink the code even further, you could set

    @blocks = { stdout => @stdout_block, stderr => @stderr_block }

(in which case you wouldn't even need `@stdout_block` and `@stderr_block` anymore, but you get the idea) then this whole processing code could become

    if @blocks[fd].call(chunk)
      raise Aborted
    end